messari-cli/src/streaming_fast/process_substream.rs

```diff
@@ -0,0 +1,303 @@
+    } else if block_input_types.len() > 1 {
+        panic!("Couldn't determine default chain from block type! More than one module input type ending in \".Block\" was specified for this substream leading to too much ambiguity for deciding which chain to pick!");
+    }
+
+    let block_type = block_input_types.into_iter().next().unwrap();
+
+    Chain::default_for_block_type(&block_type)
+}
+
+/// Returns block range info in the form -> (start_block_num, stop_block_num)
+pub(crate) async fn get_block_range_info(spkg: Vec<u8>, module_name: &str, location_type: LocationType, data_location_path: Option<PathBuf>, start_block_override: Option<i64>, chain_override: Option<Chain>) -> (i64, i64) {
+    let package = Package::decode(spkg.as_slice()).unwrap();
+    let chain = chain_override.unwrap_or(get_chain_info(&package));
+    let (sink, proto_type_name) = get_sink_and_proto_type_name(&package, module_name, EncodingType::Parquet, location_type, data_location_path, &chain);
+
+    get_block_range(&sink, &package, &proto_type_name, &chain, start_block_override, None).await
+}
+
+async fn get_start_block_num(location: Location, package: &Package, proto_type_name: &str) -> i64 {
+    // First we check to see if we have processed data for this substream/module combination before.
+    // If we have we will take this as the starting block_number
+    let processed_block_files = match location {
+        Location::DataWarehouse(path) => {
+            let bucket_name = "data-warehouse-load-427049689281-dev";
+            let region = "us-west-2".parse().unwrap();
+            let credentials = Credentials::default().unwrap();
+            let bucket = Bucket::new(bucket_name, region, credentials).unwrap();
+            let list_response = bucket.list(path.to_string_lossy().to_string(), None).await.unwrap();
+            list_response.into_iter().map(|x| x.name).collect::<Vec<_>>()
+        }
+        Location::Local(path) => {
+            fs::read_dir(path).unwrap().into_iter().map(|path| path.unwrap().path().display().to_string()).collect::<Vec<_>>()
+        }
+    };
+
+    if processed_block_files.len() > 0 {
+        // For now we will just assume all files will be in form -> startBlock_stopBlock.fileExtension
+        let mut last_block_num_iterator = processed_block_files.into_iter().map(|file| file.split('.').next().unwrap().split('_').skip(1).next().unwrap().parse::<i64>().unwrap());
+        let mut latest_block_num = last_block_num_iterator.next().unwrap();
+        for block_num in last_block_num_iterator {
+            if block_num > latest_block_num {
+                latest_block_num = block_num;
+            }
+        }
+        return latest_block_num;
+    }
+
+    // Otherwise we will fall back to taking the initial block number specified for the given module
+    for module in package.modules.as_ref().unwrap().modules.iter() {
+        if module.output.as_ref().unwrap().r#type.as_str() == proto_type_name {
```

Comment:
This unwrap seems to fail when there is a store module in the manifest (since stores don't have output)